[NEXT-19365] Added WebP support for saving thumbnails in ThumbnailService.php

SimonNitzsche opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfil our contribution guideline (https://docs.shopware.com/en/shopware-platform-dev-en/contribution/contribution-guideline?category=shopware-platform-dev-en/contribution).

Do your changes need to be mentioned in the documentation?
Add notes on your change right now in the documentation files in /src/Docs/Resources and add them to the pull request as well. 
-->

### 1. Why is this change necessary?
When uploading an webp image to Shopware 6 the thumbnails are created as an empty file due to the ThumbnailService missing the case for webp.

### 2. What does this change do, exactly?
Instead of writing the thumbnails as empty files to the disk it will actually write their content.

### 3. Describe each step to reproduce the issue or behaviour.
1. Assign an .webp image to a category.
 2. The thumbnails cannot be displayed in the Storefront due to it being empty files.

### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

SimonNitzsche committed:
Added WebP support for thumbnails in ThumbnailService.php

When uploading an webp image to Shopware 6 the thumbnails are created as an empty file due to the ThumbnailService missing the case for webp.

SimonNitzsche committed:
Fixed typo from previous commit

SimonNitzsche committed:
Added Changelog

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/platform?pullRequest=2239) <br/>All committers have signed the CLA.

shyim commented on src/Core/Content/Media/Thumbnail/ThumbnailService.php:
```diff
@@ -462,6 +462,10 @@ private function writeThumbnail($thumbnail, MediaEntity $media, string $url, int
 
+                break;
+            case 'image/webp':
+                imagewebp($thumbnail, null, $quality);
```
Please add a functions_exists with an exception when it is missing. webp needs to be compiled

SimonNitzsche replied:
Alright, will do.

SimonNitzsche replied:
Added function_exists check

SimonNitzsche committed:
Make sure function imagewebp exists

Make sure function imagewebp exists, otherwise let Shopware know that it couldn't save the thumbnail.

mynameisbogdan commented on src/Core/Content/Media/Thumbnail/ThumbnailService.php:
```diff
@@ -462,6 +462,15 @@ private function writeThumbnail($thumbnail, MediaEntity $media, string $url, int
 
+                break;
+            case 'image/webp':
+                if (function_exists('imagewebp')) {
+                    imagewebp($thumbnail, null, $quality);
+                }
+                else {
+                    throw new ThumbnailCouldNotBeSavedException($url);
+                }
```
If you don't mind, can I make a suggestion here to remove the unnecessary `else` statement?
```suggestion
                if (!function_exists('imagewebp')) {
                    throw new ThumbnailCouldNotBeSavedException($url);
                }

               imagewebp($thumbnail, null, $quality);
```

SimonNitzsche replied:
Sure

SimonNitzsche committed:
Removal of else statement for simplification

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

SimonNitzsche committed:
Added Jira ticket number to changelog.

SimonNitzsche committed:
Adjustments to comply with PHP analysis

```
1) src/Core/Content/Media/Thumbnail/ThumbnailService.php

    ---------- begin diff ----------
@@ -464,12 +464,12 @@

                 break;
             case 'image/webp':
-                if (!function_exists('imagewebp')) {
+                if (!\function_exists('imagewebp')) {
                     throw new ThumbnailCouldNotBeSavedException($url);
                 }

                imagewebp($thumbnail, null, $quality);
-                
+
                 break;
         }
         $imageFile = ob_get_contents();
    ----------- end diff -----------


Applied checkers:

 * PhpCsFixer\Fixer\FunctionNotation\NativeFunctionInvocationFixer
 * PhpCsFixer\Fixer\Whitespace\NoWhitespaceInBlankLineFixer
```

raknison commented:
Hey @SimonNitzsche thanks for your contribution! 🎉 💙
